README.md
# HexGrid

Hexagonal grid classes in JavaScript for HTML5 games, based on the sample code at [Red Blob Games](https://www.redblobgames.com/grids/hexagons/implementation.html#third-party).

(I was having trouble getting other packages to work, so I'm doing [most of] it myself!)

## License

This project is licensed under the MIT License; see [LICENSE.md](./LICENSE.md).

---

(c) 2023 Giancarlo Whitaker


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

* TBD

## 0.0.2 - 2023-08-22

### Added

* Add comments to HexDraw.js
* Add a function, drawCircle, to scripts.js to test drawing a circle on a specific hex.

## 0.0.1 - 2023-08-21

### Added

* Add HexDraw.js file, copied from [redblobgames.com](https://www.redblobgames.com/grids/hexagons/implementation.js?2023-04-18-20-36-24).  Changes were made to support its use with my existing functions.  In particular:
    * I deleted all invocations of drawGrid, i.e. from line 232 on down.
    * I imported my Hex classes from HexGrid.js
    * I labeled each function as exportable.
* The original file notes it is licensed under the [Apache License v2.0](http://www.apache.org/licenses/LICENSE-2.0.html).  As such, I shall note it was Copyright 2015 [Red Blob Games](redblobgames@gmail.com).

## 0.0.0 - 2023-08-21

### Added

* Add class implementations
* Add testing files


src/HexGrid.js
/* Cubic-coordinate data structure and algiorithms */
export class Hex {
	constructor(q, r, s) {
		this.q = q;
		this.r = r;
		this.s = s;
	}

	// Test that the current Hex is equal to Hex b
	equals(b) {
		return this.q == b.q && this.r == b.r && this.s == b.s;
	}

	/* Add two hexes together
	 * The second hex can represent steps, i.e. take a step of Hex(2, 0, -2)
	 */
	addHex(b) {
		return new Hex(this.q + b.q, this.r + b.r, this.s + b.s);
	}

	/* Subtract one hex from this
	 * The second hex can represent steps, i.e. take a step of Hex(2, 0, -2)
	 */
	subtractHex(b) {
		return new Hex(this.q - b.q, this.r - b.r, this.s - b.s);
	}

	/* Scale this
	 */
	scaleHex(/*integer*/k) {
		return new Hex(this.q * k, this.r * k, this.s * k);
	}

	hexLength() {
		return ((Math.abs(this.q) + Math.abs(this.r) + Math.abs(this.s)) / 2);
	}

	/* hexDistance
	 * Find the distance between this and Hex b
	 */
	hexDistance(b) {
		dist = this.subtractHex(b).hexLength();
		if (dist >= 0) {
			return Math.floor(dist);
		}
		else {
			return Math.ceil(dist);
		}
	}

	direction(d) {
		return Hex.directions[d];
	}

	neighbor(direction) {
		return this.addHex(Hex.directions(direction));
	}

	diagonalNeighbor(direction) {
  	return this.add(Hex.diagonals[direction]);
  }
}

Hex.directions = [new Hex(1, 0, -1), new Hex(1, -1, 0), new Hex(0, -1, 1), new Hex(-1, 0, 1), new Hex(-1, 1, 0), new Hex(0, 1, -1)];
Hex.diagonals = [new Hex(2, -1, -1), new Hex(1, -2, 1), new Hex(-1, -1, 2), new Hex(-2, 1, 1), new Hex(-1, 2, -1), new Hex(1, 1, -2)];


/* Point class
 * Simple (X, Y) point helper class
 */
export class Point {
		constructor(x, y) {
				this.x = x;
				this.y = y;
		}
}


/* Orientation class
 * Layout helper class that stores whether the display grid is pointy (point up)
 * or flat (flat edge up)
 */
export class Orientation {
	constructor(f0, f1, f2, f3, b0, b1, b2, b3, start_angle) {
		this.f0 = f0;
		this.f1 = f1;
		this.f2 = f2;
		this.f3 = f3;
		this.b0 = b0;
		this.b1 = b1;
		this.b2 = b2;
		this.b3 = b3;
		this.start_angle = start_angle;
	}
}


/* Layout class
 * Convert between hex coordinates and screen coordinates
 * Constructor params:
 *     orientation: An Orientation object, either Layout.pointy or Layout.flat
 *     size: A Point object with the X and Y dimensions of the Hexes, in pixels
 *     origin: A Point object that is the X and Y of the origin of the Layout
 */
export class Layout {
	constructor(orientation, size, origin) {
		this.orientation = orientation;
		this.size = size;
		this.origin = origin;
	}
	hexToPixel(h) {
		var orient = this.orientation;
		var size = this.size;
		var origin = this.origin;
		var x = (orient.f0 * h.q + orient.f1 * h.r) * size.x;
		var y = (orient.f2 * h.q + orient.f3 * h.r) * size.y;
		return new Point(x + origin.x, y + origin.y);
	}
	pixelToHex(p) {
		var orient = this.orientation;
		var size = this.size;
		var origin = this.origin;
		var pt = new Point((p.x - origin.x) / size.x, (p.y - origin.y) / size.y);
		var q = orient.b0 * pt.x + orient.b1 * pt.y;
		var r = orient.b2 * pt.x + orient.b3 * pt.y;
		return new Hex(q, r, -q - r);
	}
	hexCornerOffset(corner) {
		var orient = this.orientation;
		var size = this.size;
		var angle = 2.0 * Math.PI * (orient.start_angle - corner) / 6.0;
		return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle));
	}
	polygonCorners(h) {
		var corners = [];
		var center = this.hexToPixel(h);
		for (var i = 0; i < 6; i++) {
			var offset = this.hexCornerOffset(i);
			corners.push(new Point(center.x + offset.x, center.y + offset.y));
		}
		return corners;
	}
}

Layout.pointy = new Orientation(Math.sqrt(3.0), Math.sqrt(3.0) / 2.0, 0.0, 3.0 / 2.0, Math.sqrt(3.0) / 3.0, -1.0 / 3.0, 0.0, 2.0 / 3.0, 0.5);
Layout.flat = new Orientation(3.0 / 2.0, 0.0, Math.sqrt(3.0) / 2.0, Math.sqrt(3.0), 2.0 / 3.0, 0.0, -1.0 / 3.0, Math.sqrt(3.0) / 3.0, 0.0);


src/scripts.js
import { Hex } from "./HexGrid.js"
import { Point } from "./HexGrid.js"
import { Orientation} from "./HexGrid.js"
import { Layout } from "./HexGrid.js"
import * as HexDraw from "./HexDraw.js"

const canvas = document.getElementById("grid");
const ctx = canvas.getContext("2d");

let h = HexDraw.shapeHexagon(6);
let l = new Layout(Layout.pointy, new Point(25, 25), new Point(0, 0));
console.log(h);
HexDraw.drawGrid({id: "grid", labels: true,
          layout: l, hexes: h})

// Add a custom object
// A Boolean to draw a circle in the hex
// Add to hex (4, -2, -2)

// -3, 6, -3
h[107].drawCircle = true;

function drawCircle (layout, hex, ctx) {
  let center = layout.hexToPixel(hex);
  if (hex.drawCircle != null) {
    console.log("Not null!", hex);
    ctx.beginPath();
    // arc(x, y, radius, startAngle, endAngle)
    console.log(center.x, ", ", center.y);
    ctx.arc(250+center.x, 250+center.y, 15, 0, 2 * Math.PI);
    ctx.fillStyle = "red";
    ctx.fill();
    ctx.closePath();
  }
}

for (let i = 0; i < h.length; i++) {
  drawCircle(l, h[i], ctx);
}


src/HexDraw.js
// From http://www.redblobgames.com/grids/hexagons/implementation.html
// Copyright 2015 Red Blob Games <redblobgames@gmail.com>
// License: Apache v2.0 <http://www.apache.org/licenses/LICENSE-2.0.html>

// This code is used to generate the diagrams on implementation.html

// Edits by Giancarlo Whitaker, 2023

/* global Hex, Layout, Point */
import { Hex } from "./HexGrid.js"
import { Point } from "./HexGrid.js"
import { Orientation} from "./HexGrid.js"
import { Layout } from "./HexGrid.js"


/* drawHex
 * Draw a Hex object to the screen
 * param ctx: <canvas> drawing context (must be 2d)
 * param layout: a Layout object
 * param hex: a Hex object
 * param style{}:
 */
export function drawHex(ctx, layout, hex, style={}) {
    var corners = layout.polygonCorners(hex);
    ctx.beginPath();
    ctx.strokeStyle = "black";
    ctx.fillStyle = "white";
    ctx.lineWidth = 1;
    Object.assign(ctx, style);
    ctx.moveTo(corners[5].x, corners[5].y);
    for (var i = 0; i < 6; i++) {
        ctx.lineTo(corners[i].x, corners[i].y);
    }
    ctx.fill();
    ctx.stroke();
}

/* colorForHex
 * Find stylings to the Hex on the screen
 * param hex: a Hex object
 */
export function colorForHex(hex) {
    // Match the color style used in the main article
    if (hex.q === 0 && hex.r === 0 && hex.s === 0) {
        return "hsl(0, 50%, 0%)";
    } else if (hex.q === 0) {
        return "hsl(90, 70%, 35%)";
    } else if (hex.r === 0) {
        return "hsl(200, 100%, 35%)";
    } else if (hex.s === 0) {
        return "hsl(300, 40%, 50%)";
    } else {
        return "hsl(0, 0%, 50%)";
    }
}

/* drawHexLabel
 * Apply a label to the Hex on the screen
 * param ctx: <canvas> drawing context (must be 2d)
 * param layout: a Layout object
 * param hex: a Hex object
 * param style{}:
 */
export function drawHexLabel(ctx, layout, hex, style={}) {
    const pointSize = Math.round(0.5 * Math.min(Math.abs(layout.size.x), Math.abs(layout.size.y)));
    var center = layout.hexToPixel(hex);
    ctx.fillStyle = colorForHex(hex);
    ctx.font = `${pointSize}px sans-serif`;
    ctx.textAlign = "center";
    ctx.textBaseline = "middle";
    Object.assign(ctx, style);
    ctx.fillText(hex.hexLength() === 0? "q,r,s" : (hex.q + "," + hex.r + "," + hex.s), center.x, center.y);
}

export function permuteQRS(q, r, s) { return new Hex(q, r, s); }
export function permuteSRQ(s, r, q) { return new Hex(q, r, s); }
export function permuteSQR(s, q, r) { return new Hex(q, r, s); }
export function permuteRQS(r, q, s) { return new Hex(q, r, s); }
export function permuteRSQ(r, s, q) { return new Hex(q, r, s); }
export function permuteQSR(q, s, r) { return new Hex(q, r, s); }



/*
 *
 * MAP FUNCTIONS
 *
 */
export function shapeParallelogram(q1, r1, q2, r2, constructor) {
    var hexes = [];
    for (var q = q1; q <= q2; q++) {
        for (var r = r1; r <= r2; r++) {
            hexes.push(constructor(q, r, -q-r));
        }
    }
    return hexes;
}

export function shapeTriangle1(size) {
    var hexes = [];
    for (var q = 0; q <= size; q++) {
        for (var r = 0; r <= size-q; r++) {
            hexes.push(new Hex(q, r, -q-r));
        }
    }
    return hexes;
}

export function shapeTriangle2(size) {
    var hexes = [];
    for (var q = 0; q <= size; q++) {
        for (var r = size-q; r <= size; r++) {
            hexes.push(new Hex(q, r, -q-r));
        }
    }
    return hexes;
}

export function shapeHexagon(size) {
    var hexes = [];
    for (var q = -size; q <= size; q++) {
        var r1 = Math.max(-size, -q-size);
        var r2 = Math.min(size, -q+size);
        for (var r = r1; r <= r2; r++) {
            hexes.push(new Hex(q, r, -q-r));
        }
    }
    return hexes;
}

export function shapeRectanglePointy(left, top, right, bottom) {
    let hexes = [];
    for (let r = top; r <= bottom; r++) {
        let r_offset = Math.floor(r/2.0); // or r>>1
        for (let q = left - r_offset; q <= right - r_offset; q++) {
            hexes.push(new Hex(q, r, -q-r));
        }
    }
    return hexes;
}

export function shapeRectangleFlat(left, top, right, bottom) {
    let hexes = [];
    for (let q = left; q <= right; q++) {
        let q_offset = Math.floor(q/2.0); // or q>>1
        for (let r = top - q_offset; r <= bottom - q_offset; r++) {
            hexes.push(new Hex(q, r, -q-r));
        }
    }
    return hexes;
}

export function shapeRectangleArbitrary(w, h, constructor) {
    var hexes = [];
    var i1 = -Math.floor(w/2), i2 = i1 + w;
    var j1 = -Math.floor(h/2), j2 = j1 + h;
    for (var j = j1; j < j2; j++) {
        var jOffset = -Math.floor(j/2);
        for (var i = i1 + jOffset; i < i2 + jOffset; i++) {
            hexes.push(constructor(i, j, -i-j));
        }
    }
    return hexes;
}

/* drawGrid
 * Draw a grid of Hexes onto the screen
 * param id: the id field for the canvas, i.e. <canvas id="param">
 * param labels: Boolean that is true if labels are to be drawn
 * param layout: a Layout object
 * param hexes: a map of Hex objects, i.e. a Hex data structure
 * param xMarkers, yMarkers: I (Giancarlo) don't know what these do.  They're optional!
 */
export function drawGrid({id, labels, layout, hexes, axes, xMarkers, yMarkers}) {
    const margin = axes ? {left: 30, right: 2, top: 2, bottom: 20} : {left: 0, right: 0, top: 0, bottom: 0};
    labels = labels ?? false;
    hexes = hexes ?? shapeRectangleArbitrary(15, 15, permuteQRS);

    const canvas = /** @type{HTMLCanvasElement} */(document.getElementById(id));
    if (!canvas) { console.warn(`Could not find canvas id=${id}`); return; }
    const ctx = canvas.getContext('2d');

    const width = canvas.width;
    const height = canvas.height;
    const innerWidth = width - margin.left - margin.right;
    const innerHeight = height - margin.bottom - margin.top;
    if (window.devicePixelRatio && window.devicePixelRatio != 1) {
        canvas.width = width * window.devicePixelRatio;
        canvas.height = height * window.devicePixelRatio;
        ctx.scale(window.devicePixelRatio, window.devicePixelRatio);
    }

    ctx.clearRect(0, 0, width, height);

    // Draw the axes
    if (axes) {
        const gridSpacing = 10;
        ctx.save();
        ctx.font = "10px sans-serif";
        ctx.fillStyle = "hsl(180 30% 30%)"
        ctx.translate(margin.left + innerWidth/2, margin.top + innerHeight/2);

        ctx.textAlign = "right";
        const firstRow = Math.ceil(-innerHeight/2 / gridSpacing),
              lastRow = Math.floor(innerHeight/2 / gridSpacing);
        for (let r = firstRow; r <= lastRow; r++) {
            const y = gridSpacing * r;
            const major = r % 5 === 0;
            ctx.strokeStyle = major? "hsl(180 30% 50%)" : "hsl(180 30% 80%)";
            ctx.beginPath();
            ctx.moveTo(-innerWidth/2 - (major ? 7 : 5), y);
            ctx.lineTo(innerWidth/2, y);
            ctx.stroke();
            if (major) ctx.fillText(r * gridSpacing, -innerWidth/2- 7, y + 3);
        }

        ctx.textAlign = "center";
        const firstColumn = Math.ceil(-innerWidth/2 / gridSpacing),
              lastColumn = Math.floor(innerWidth/2 / gridSpacing);
        for (let q = firstColumn; q <= lastColumn; q++) {
            const x = gridSpacing * q;
            const major = q % 5 === 0;
            ctx.strokeStyle = major? "hsl(180 30% 50%)" : "hsl(180 30% 80%)";
            ctx.beginPath();
            ctx.moveTo(x, -innerHeight/2);
            ctx.lineTo(x, innerHeight/2 + (major ? 7 : 5));
            ctx.stroke();
            if (major) ctx.fillText(q * gridSpacing, x, innerHeight/2 + 15);
        }
        ctx.restore();
    }

    // Draw the hexagons
    ctx.save();
    const clipArea = new Path2D();
    clipArea.rect(margin.left, margin.top, innerWidth, innerHeight);
    ctx.clip(clipArea);
    ctx.translate(margin.left + innerWidth/2, margin.top + innerHeight/2);
    hexes.forEach(function(hex) {
        drawHex(ctx, layout, hex);
        if (labels) drawHexLabel(ctx, layout, hex);
    });
    ctx.restore();

    // Draw any horizontal/vertical lines
    ctx.save();
    ctx.beginPath();
    ctx.lineWidth = 1.5;
    ctx.strokeStyle = "hsl(180 30% 30% / 0.5)";
    ctx.setLineDash([3, 5]);
    for (let x of xMarkers ?? []) {
        ctx.moveTo(margin.left + innerWidth/2 + x, margin.top);
        ctx.lineTo(margin.left + innerWidth/2 + x, margin.top + innerHeight);
    }
    for (let y of yMarkers ?? []) {
        ctx.moveTo(margin.left, margin.top + innerHeight/2 + y);
        ctx.lineTo(margin.left + innerWidth, margin.top + innerHeight/2 + y);
    }
    ctx.stroke();
    ctx.restore();
}
